### ISSUE TITLE
Proxito: standardize how to handle redirects from subprojects

### ISSUE THREAD

**Issue by stsewd** [readthedocs/readthedocs.org#10049]
Currently, we don't have tests nor documentation for this.

We do check the redirects from subprojects when a matching `/projects/subproject/` path 404.
The current behavior is as follows:

### Exact redirects

We try to match the subproject redirects using the full path `/projects/subproject/foo.html`, we don't check for the redirects of the parent project. The generated redirect will be absolute to the current domain.

For example, the following redirect defined in the subproject

- from: /projects/subproject/foo.html
- to: /en/latest/

Will redirect `docs.example.com/projects/subproject/foo.html` to `docs.example.com/en/latest/`

### Page redirect

We try to match the redirects using the captured filename `foo.html`, we don't check for the redirects of the parent project. The generated redirect will be relative to the subproject.

For example, the following redirect defined in the subproject

- from: foo.html
- to: /

Will redirect `docs.example.com/projects/subproject/en/latest/foo.html` to `docs.example.com/projects/subproject/en/latest/`.

### Prefix redirect

We try to match the redirects using the captured filename `foo.html`, we don't check for the redirects of the parent project. The generated redirect will be relative to the subproject.

For example, the following redirect defined in the subproject

- from: /prefix/

Will redirect `docs.example.com/projects/subproject/prefix/foo.html` to `docs.example.com/projects/subproject/en/latest/foo.html`.

### Changes

I think page and prefix redirect are ok, exact redirects need to change to a less confusing implementation. Something like:

Match the path after the `/projects/subproject/` part, the generated redirect will still be absolute to the current domain.

For example, the following redirect defined in the subproject

- from: /foo.html
- to: /en/latest/

Will redirect `docs.example.com/projects/subproject/foo.html` to `docs.example.com/en/latest/`.

**Comment by humitos** [comment]
IMO, subproject redirects shold _always relative_ to the path prefix. If users want to redirect _outside_ the subproject, they should use a full URL for the `to` argument. I think that's the easiest way to explain.

With that in mind,
* Type: Exact
* From: `/foo.html`
* To: `/new/bar.html`

Will redirect:
* `https://docs.example.com/projects/subproject/foo.html`
* `https://docs.example.com/projects/subproject/new/bar.html`

If the user wants to go _outsite_ the subroject, they would need to write the full URL as:
* Type: Exact
* From: `/foo.html`
* To: `https://docs.example.com/en/latest/bar.html`


Following this idea, _all the redirects_ will be inside their own projects unless explicitly set via a full URL it has to go outsite it. Besides, this logic follows the current implementation for the other 2 types: Page and Prefix.

Extra points: if we _really want to go outside the subproject_ without writing a full URL, we should have an extra argument/field for these exact redirect where we can "force" to start considering the path from `/`. The previous example would be:
* Type: Exact
* From: `/foo.html`
* To: `/new/bar.html`
* Force from `/`: True

Will redirect:
* `https://docs.example.com/projects/subproject/foo.html`
* `https://docs.example.com/new/bar.html`


This allows subprojects to be "self-contained" and move them from being a subproject to main project and keep their redirects always working.

**Comment by ericholscher** [comment]
I'm still -1 on exact redirects not matching the exact URL.

> With that in mind,
> 
> * Type: Exact
> * From: `/foo.html`
> * To: `/new/bar.html`
> 
> Will redirect:
> 
> * `https://docs.example.com/projects/subproject/foo.html`
> * `https://docs.example.com/projects/subproject/new/bar.html`

Isn't this what `Page` redirects are for? 

> If the user wants to go _outsite_ the subroject, they would need to write the full URL as:
> 
> * Type: Exact
> * From: `/foo.html`
> * To: `https://docs.example.com/en/latest/new/bar.html`
 
I think this also makes sense. I think saying `Exact redirects only match inside the subproject, but can redirect to any fully formed URL` seems reasonable? Having them work differently between parent projects and subprojects feels *really* confusing.

I think they should work the same (match the whole URL path), but have restrictions on when they are matched.

> Following this idea, _all the redirects_ will be inside their own projects unless explicitly set via a full URL it has to go outsite it. Besides, this logic follows the current implementation for the other 2 types: Page and Prefix.

Right, but those are defined as understanding versions/projects, and not the exact path (it's in the name! Exact Redirect). I'm not saying Exact Redirects are the right abstraction, but if we want another solution, we should make it instead of changing how Exact Redirects work sometimes.

> Extra points: if we _really want to go outside the subproject_ without writing a full URL, we should have an extra argument/field for these exact redirect where we can "force" to start considering the path from `/`. The previous example would be:
> 
> * Type: Exact
> * From: `/foo.html`
> * To: `/new/bar.html`
> * Force from `/`: True
> 
> Will redirect:
> 
> * `https://docs.example.com/projects/subproject/foo.html`
> * `https://docs.example.com/new/bar.html`
> 
> This allows subprojects to be "self-contained" and move them from being a subproject to main project and keep their redirects always working.

This is adding more edge cases and confusion to the user, I think. I don't think this is the right abstraction for additional work on redirects, since I think Redirects v2 should likely remove all these "redirect type" concepts and just work with regex or similar as we've discussed in the past.

**Comment by humitos** [comment]
@ericholscher 

> My thinking is still: Redirects work the same in subprojects as in normal projects, but they are restricted in the path they can be created/evaluated on.

👍🏼 -- can you add some detailed examples about how this implementation would work exactly? I think that will help us to be all in the same page here and avoid miss-understandings.

**Comment by ericholscher** [comment]
Basically:

## Subproject redirect

A subproject that is served on `docs.example.com/projects/subproject/en/latest/`

* Type: Exact
* From: /projects/subproject/foo/
* To: /projects/subproject/bar/
* Result: This will work and redirect `docs.example.com/projects/subproject/foo/` to `docs.example.com/projects/subproject/bar/`

* Type: Exact
* From: /projects/subproject/foo/
* To: https://docs.example.com/projects/subproject2/en/latest/ (or any other fully formed URL)
* Result:  This works and redirects 1 subproject to another

* Type: Exact
* From: /foo/
* To: /bar/
* Result: Form validation error, subproject redirects must start with `/projects/subproject/`

## Parent project redirect

* From: /projects/subproject/foo/
* To: /projects/subproject/bar/
* Result: Form validation error, subproject redirects must start be set on the subproject

* From: /foo
* To: /projects/subproject/bar/
* Result: This will work and redirect `docs.example.com/foo/` to `docs.example.com/projects/subproject/bar/`

**Comment by stsewd** [comment]
I think it could also be helpful if we match the exact redirects from the parent project, that way users can have all redirects in one central place, just an idea. For example, translations, subprojects, translations from subprojects, etc, all of those paths will check the exact redirects from the project that owns the domain (parent project) in addition to the redirects from the current project.

And maybe force to always have exact redirects in the parent project by disabling exact redirects from the translations/subprojects.

**Comment by humitos** [comment]
@ericholscher would this work?

* Type: Exact
* From: `/projects/subproject/foo/`
* To: `/projects/another-subproject/foo/`
* Result:  This works and redirects 1 subproject to another

I guess it would. It's similar to your full URL example, but using a relative one.

----

I think your proposal makes everything pretty explicit and seems easy to explain to users. Also, validation errors make more sense and there is nothing hidden to users (as I was proposing with the "hidden scope"), 👍🏼

**Comment by humitos** [comment]
@stsewd 

> I think it could also be helpful if we match the exact redirects from the parent project, that way users can have all redirects in one central place, just an idea.

Hrm, isn't this opposed to what @ericholscher is suggesting with the form validation errors where "subproject redirects must be set on the subproject"?

**Comment by humitos** [comment]
@stsewd yeah, that's the trade off of these solutions. The more explicit we make it for users, the less magic we can apply behind the scenes. However, if people want to change the alias and fix a bunch of redirects, they can always use the API for that

**Comment by ericholscher** [comment]
I think I'm also OK with the parent project being able to set redirects on the subproject path. I agree it makes it easier to keep everything in one place -- I was a bit conflicted on which approach made sense in my above comment, so happy with either.

The main thing I want is to ensure subproject redirects only operate on the subproject namespace. That seems like the most important thing to me.

**Comment by agjohnson** [comment]
I had reports of users confused by our current implementation, I feel like this could be worth fixing.

I agree with what @humitos and @stsewd were saying above early on, exact redirects make more sense relative to the subproject. Especially with superprojects able to define the subproject alias now, it feels like bad design to bake the subproject URL into the subproject's redirect if that can change outside the subproject.

Setting these redirects at the superproject doesn't seem great either because that pushes management for a project outside that project -- especially a problem on business where a team managing a subproject could not have access to the superproject.

On the naming, "exact" is confusing and we could get away from the term driving design. These redirects are already not exact (they glob match wildcard) and given what most of us mentioned you can certainly interpret "exact, relative to the subproject URL". It seems like users interpret this simply as a "the redirect that is not page redirect" though, which I think is very fair. This type would probably be clearer not called "exact" at all.

Exact redirects are needed to use wildcards so page redirects aren't useful here. The case I'm describing is: exact redirect with force redirect on a version `/en/2.2/*` to `/en/2.x/:splat`.

I put up

- #13302

It supports both patterns too, so we could even punt on this all. There are 123 redirects using the superproject path workaround currently.

**Comment by humitos** [comment]
I agree with you @agjohnson here and, based on the firsts comments from @ericholscher, it seems that the name `Exact redirects` is what carries the confusion here.

IMHO, redirects on subprojects should be relative to the subproject path. It should not take into consideration the "base path" (`/projects/<alias>/`) for the reasons that Anthony mentioned: the base path could change, the user doesn't have permissions on the super project to add a redirect, etc. This is the implementation I'd like to have.

If we think the name is wrong and we can't change it for any reason, we should create another type of redirect -- but I'd avoid that if it's possible. However, the "subproject redirect based on the relative path" should exist somehow.

**Comment by MTL1010000101** [comment]
> Currently, we don't have tests nor documentation for this.
> 
> We do check the redirects from subprojects when a matching `/projects/subproject/` path 404.
> The current behavior is as follows:
> 
> ### Exact redirects
> 
> We try to match the subproject redirects using the full path `/projects/subproject/foo.html`, we don't check for the redirects of the parent project. The generated redirect will be absolute to the current domain.
> 
> For example, the following redirect defined in the subproject
> 
> - from: /projects/subproject/foo.html
> - to: /en/latest/
> 
> Will redirect `docs.example.com/projects/subproject/foo.html` to `docs.example.com/en/latest/`
> 
> ### Page redirect
> 
> We try to match the redirects using the captured filename `foo.html`, we don't check for the redirects of the parent project. The generated redirect will be relative to the subproject.
> 
> For example, the following redirect defined in the subproject
> 
> - from: foo.html
> - to: /
> 
> Will redirect `docs.example.com/projects/subproject/en/latest/foo.html` to `docs.example.com/projects/subproject/en/latest/`.
> 
> ### Prefix redirect
> 
> We try to match the redirects using the captured filename `foo.html`, we don't check for the redirects of the parent project. The generated redirect will be relative to the subproject.
> 
> For example, the following redirect defined in the subproject
> 
> - from: /prefix/
> 
> Will redirect `docs.example.com/projects/subproject/prefix/foo.html` to `docs.example.com/projects/subproject/en/latest/foo.html`.
> 
> ### Changes
> 
> I think page and prefix redirect are ok, exact redirects need to change to a less confusing implementation. Something like:
> 
> Match the path after the `/projects/subproject/` part, the generated redirect will still be absolute to the current domain.
> 
> For example, the following redirect defined in the subproject
> 
> - from: /foo.html
> - to: /en/latest/
> 
> Will redirect `docs.example.com/projects/subproject/foo.html` to `docs.example.com/en/latest/`.
> 
> I think we should be fine ignoring redirects from the parent if we are handling a 404 from a subproject.
> 
> <img src="https://front.com/assets/img/favicons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_gq71eq)

**Comment by agjohnson** [comment]
Naming feels like a slight source of confusion, but I think we could change this behavior of subproject redirects and users wouldn't be confused by the change, or be confused by the naming at least. I think this is a case where the naming probably has a more significant meaning to us because we know the technical implementation and history.

I did have a similar thought on a third redirect type initially, something like page, absolute, relative redirects. But that would make the differences between the three harder to understand.

I don't know exactly what I would suggest long term, but I do sort of want just one redirect type with an option for "redirect this URL for each version" to signal a page redirect. There is more graceful phrasing there, I haven't thought about that seriously. But the point would be to avoid putting page and exact redirects at odds. This is a separate conversation of course though.

**Comment by ericholscher** [comment]
Can we just support the redirect working if it's made on either the parent or subproject? Is that the primary confusion?

**Comment by agjohnson** [comment]
The main point of confusion is requiring the subproject prefix in the redirect for it to work. I think _where_ the redirect lives is only confusing given this conversation, it's not something I think users would expect naturally -- this pattern of setting the subproject redirect on the superproject feels more awkward than requiring a prefix.

**Comment by MTL1010000101** [comment]
Why confused ME managing director MOMIN TRADING LTD. MD. MOKHLESUR RAHMAN MTL 1010000101
BEST REGARDS 🩵

**Comment by stsewd** [comment]
After checking https://github.com/readthedocs/readthedocs.org/pull/13302 I think I'm on favor of:

- Exact redirects on subprojects should use the relative path for the redirect (omit the /projects/subproject prefix).
- Remove the prefix from existing redirects, so they keep working.
- Decide if we should remove redirects that weren't working before, but will start working now.
- Communicate this change to users, so they aren't surprised about the change in behavior (this is probably a small set of users given the few redirects we found).

Having the redirects apply to the path without the prefix, allows users to change the subproject prefix without problems, or to add an existing project as a subproject of another without having to change the current redirects.

**Comment by agjohnson** [comment]
I feel about the same but I'll add that I'm also fine shipping what is in #13302 currently to start -- support for both relative URLs and prefixed URLs simultaneously. From there we can work backwards towards deprecation like you're describing

This gives us a immediate fix but also gives users a window to make the change to a relative URL themselves. If we just change the behavior to use relative URLs, dropping support for URLs with a prefix, we have to coordinate a change to all existing redirects during the release. The code added to support already prefixed redirect URLs was not complicated at least.

If we don't feel good about the extra magic we could also add an explicit boolean field "relative URL" for any existing redirect and decide how that works in to deprecation or default behavior.

**Comment by humitos** [comment]
I agree with both here. I'd say that the immediate change to support relative URL is good and unblock the users. However, the end goal here is what @stsewd posted in https://github.com/readthedocs/readthedocs.org/issues/10049#issuecomment-5704647900. We shouldn't deploy the immediate change as a quick fix and forget about the end goal 😉 --we should continue working on that after shipping it.

**Comment by agjohnson** [comment]
Absolutely, we can even start the deprecation process now if we want to talk about the change and state a date for full deprecation in one place. Let's track this as a longer project to make sure we have follow up steps covered.
